### DISCUSSION TITLE
filtering data provider where data is from an SQL query

### DISCUSSION THREAD

**Discussion by kmdelucaNG** [Budibase/budibase#8144] [Help] [answered]
This is on self-hosted v1.3.11.

I have a Data Provider that gets its content from a custom SQL query.  The data displays fine but none of the front-end filtering works using the _Define Filters_.
If I change the data source to an SQL Table, then the filtering works fine. 
Also tried it with the Dynamic Filter..., same behavior.

Is that expected ?

thanks.

**Comment by melohagan** [comment, marked as answer]
Hi @kmdelucaNG 

When you use custom queries, filtering will be done via front-end JavaScript, whereas SQL datasources will use SQL for filtering.

I updated the docs recently to explain this a little bit: https://docs.budibase.com/docs/data-provider#filtering

**Reply by kmdelucaNG** [reply]
thanks @melohagan.
I had read the Data Provider filtering doc.  Was there an example of filtering via JavaScript that was provided in there?
And not sure if I was clear - but I do have an SQL datasource - my custom query is postgreSQL.

**Reply by gmentzik** [reply]
It is possible to have filtering in queries if you create a custom query with where clause and like '%{{some binding}}%' statements. Then use the little gog next to Provider name at page and not filtering. However I must say that using custom queries for bit chunks of data is not handled well by Budibase as, as mentioned pagination is not supported, so you must redevelop yourself pagination funtion at FE adding a LIMIT OFFSET to your query... My experiense so far is that custom queries are more to bring back a special row or perform something very specific You cannot with ready made Budibase providers.

**Reply by melohagan** [reply]
Excellent summary @gmentzik Thanks! As mentioned it is better to do the filtering in your custom query rather than use JavaScript on the front-end.

**Reply by gmentzik** [reply]
Adding here a custom query for MySQL example I developed with the option do have empty values to help others:
```
WITH 
o as ( 
select * from operator WHERE use_as_operator = 1 
)
SELECT o.id,o.operator_network as operator,c.id as country_id,c.country_alpha3,c.country_name,c.market_area  
FROM o join country as c on c.id = o.country_id
WHERE ((ISNULL({{country_id}}) OR TRIM({{country_id}}) = '' OR {{country_id}} = 0) OR country_id = {{country_id}})
AND ((ISNULL({{operator_name}}) OR TRIM({{operator_name}}) = '') OR operator_network LIKE '%{{operator_name}}%')
ORDER BY o.id LIMIT {{limit}};
```
as said in the end I went for providers as they are more flexible than custom queries.

**Reply by agruenberg** [reply]
OT, but why are you using isnull() instead of is null?
My experience is that is null is more efficient and faster than isnull() (seek vs scan). And in my opinion it is much more idiomatic (what is my main reason using is null)

**Reply by gmentzik** [reply]
> OT, but why are you using isnull() instead of is null? My experience is that is null is more efficient and faster than isnull() (seek vs scan). And in my opinion it is much more idiomatic (what is my main reason using is null)

Probably it is. As said this is just an example to help others, as I was trying to find a way to use mysql custom queries as a provider (having empty values at filter (cog)) and could not find many examples. In the end I rejected as 3000 rows of data is too much for frontend processing and was not in favor of redeveloping pagination feature though is possible using states. With a bit of flexibility in design I was able to go for standard providers. But this may not always be the case and you may not be always be able to use standard providers.
